evalml/automl/automl_search.py

```diff
@@ -575,7 +589,15 @@ def _find_best_pipeline(self):
             if self._train_best_pipeline:
                 X_threshold_tuning = None
                 y_threshold_tuning = None
-                X_train, y_train = self.X_train, self.y_train
+                if self._best_pipeline.model_family == ModelFamily.ENSEMBLE:
+                    X_train, y_train = self.X_ensemble, self.y_ensemble
+                else:
+                    if self.X_ensemble:
+                        X_train = infer_feature_types(self.X_train.to_dataframe().append(self.X_ensemble.to_dataframe(), ignore_index=True), self.X_feature_types)
+                        y_train = infer_feature_types(self.y_train.to_series().append(self.y_ensemble.to_series(), ignore_index=True))
```

Comment:
Hmm. Converting to pandas and then calling `infer_feature_types` again means we lose of the custom user-defined types. @angela97lin had to do a bunch of plumbing to avoid this during pipeline evaluation!

An example of this (which would be a great test to add to this PR): user overrides woodwork to set a text feature as categorical. We mock out the ensemble fit method to ensure that the categorical type makes its way to the metalearner instead of the default inferred text type.

It also feels strange to split up the data in the `AutoMLSearch` constructor, and then merge it again here.

The most compelling option I could think of right now is: in `AutoMLSearch` at least, pass around indices instead of dataframes/datatables, until the moment we actually need a sub-datatable in order to proceed; then at that point, slice as needed.

We could also keep references to both a) the complete data and b) the splitted data, and use whichever we need when we need it. The difficulty there is that effectively doubles our memory consumption.

Let's discuss.

Reply:
@dsherry I see what you mean. For the custom user-defined types, wouldn't passing in `self.X_train.logical_types` to `infer_feature_types` solve that problem? 

As for the issue of splitting and merging unnecessarily, I think the best idea would then be to keep a reference to the indices of the ensembling data rather than the full split ensembling data. This would also help with memory consumption but should allow us to keep the same functionality as everything before. What do you think?

Reply:
@bchen1116 oh maybe so! @angela97lin does that line up with your understanding? I suggest you check with her on this.

> I think the best idea would then be to keep a reference to the indices of the ensembling data rather than the full split ensembling data. 

If I'm following you correctly here, yep, I'm suggesting we look into doing the splitting on indices first instead of on datatables.

If using indices is not required to fix the type problem, we could file it as a separate issue, but let's be careful about the code we use to do the concatenation here. In fact I'd suggest we use `pd.concat` instead of `append` here.

Reply:
@bchen1116 @dsherry 
RE using `infer_feature_types`: If the *only* thing that's happening is we're taking a subset of rows, then passing `self.X_train.logical_types` to `infer_feature_types` should suffice since nothing is happening to the `logical_types`.

For the transformers, we converted to pandas, and then potentially transformed the data before we converted back to Woodwork, so extra plumbing was necessary (and stored in `_retain_custom_types_and_initalize_woodwork`).

Of course, I agree that testing would be important here to make sure we're not missing anything :)


I do agree that it feels pretty strange to split up the data in the AutoMLSearch constructor, and then merge it again here though...

Reply:
Just updated the implementation to track the indices instead, so we no longer need to do any `concat/append` logic when training the best pipeline.